File: src/dataset/python/bowling_exercism-ThomasZumsteg/gemini/GeminiFlash_bowling_v4.py
class BowlingGame(object):
    def __init__(self):
        self._rolls = []
        self._frame = 0
        self._first = True
        self._extra_rolls = 0

    def roll(self, pins):
        if self._frame >= 20 and self._extra_rolls <= 0:
            raise IndexError("Game Over")

        if not (0 <= pins <= 10):
            raise ValueError("Cannot score {}".format(pins))

        if not self._first and len(self._rolls) > 0 and pins + self._rolls[-1] > 10:
            raise ValueError("Cannot score {} and {}".format(self._rolls[-1], pins))

        self._rolls.append(pins)
        self._update_state(pins)

    def score(self):
        if self._frame < 20 or self._extra_rolls > 0:
            raise IndexError("Game in progress")

        total = 0
        index = 0
        for frame in range(10):
            first = self._rolls[index]
            if first == 10:  # Strike
                second = self._rolls[index + 1] if index + 1 < len(self._rolls) else 0
                third = self._rolls[index + 2] if index + 2 < len(self._rolls) else 0
                total += 10 + second + third
                index += 1
            else:
                second = self._rolls[index + 1]
                if first + second == 10:  # Spare
                    third = self._rolls[index + 2] if index + 2 < len(self._rolls) else 0
                    total += 10 + third
                else:
                    total += first + second
                index += 2
        return total

    def _update_state(self, pins):
        strike = pins == 10

        if strike:
            self._frame += 2
            self._first = True
        else:
            self._frame += 1
            self._first = not self._first

        if self._frame == 20:
            if strike:
                self._extra_rolls = 2
            elif len(self._rolls) >= 2 and sum(self._rolls[-2:]) == 10:
                self._extra_rolls = 1
        elif self._extra_rolls > 0:
            self._extra_rolls -= 1

    def _game_over(self):
        return self._frame >= 20 and self._extra_rolls <= 0

File: src/dataset/javascript/bowling_exercism-ThomasZumsteg/gemini/GeminiFlash_bowling_v4.js
class Bowling {
    constructor() {
        this.rolls = [];
        this.frame = 0;
        this.extraRolls = 0;
        this.rollInFrame = 0;
    }

    gameOver() {
        return this.frame >= 10 && this.extraRolls <= 0;
    }

    roll(pins) {
        if (this.gameOver()) {
            throw new Error('Cannot roll after game is over');
        }

        if (pins < 0) {
            throw new Error('Negative roll is invalid');
        }

        const currentFrameRolls = this.rolls.slice(-1);
        if (pins > 10 || (this.rollInFrame === 1 && currentFrameRolls[0] + pins > 10)) {
            throw new Error('Pin count exceeds pins on the lane');
        }

        if (this.frame === 9) {
            if (this.rollInFrame === 0 && pins === 10) {
                this.extraRolls = 2;
            } else if (this.rollInFrame === 1 && currentFrameRolls[0] + pins === 10) {
                this.extraRolls = 1;
            }
        } else if (this.frame > 9) {
            this.extraRolls--;
        }

        this.rolls.push(pins);

        if (pins === 10) {
            this.frame++;
            this.rollInFrame = 0;
        } else {
            this.rollInFrame++;
            if (this.rollInFrame === 2) {
                this.frame++;
                this.rollInFrame = 0;
            }
        }
    }

    score() {
        if (!this.gameOver()) {
            throw new Error('Score cannot be taken until the end of the game');
        }

        let score = 0;
        let rollIndex = 0;

        for (let frame = 0; frame < 10; frame++) {
            const firstRoll = this.rolls[rollIndex];
            const secondRoll = this.rolls[rollIndex + 1];

            if (firstRoll === 10) { // Strike
                score += 10 + this.rolls[rollIndex + 1] + this.rolls[rollIndex + 2];
                rollIndex++;
            } else if (firstRoll + secondRoll === 10) { // Spare
                score += 10 + this.rolls[rollIndex + 2];
                score += firstRoll + secondRoll;
                rollIndex += 2;
            } else {
                score += firstRoll + secondRoll;
                rollIndex += 2;
            }
        }

        return score;
    }
}

module.exports = Bowling;